src/erasure-code/clay/ErasureCodeClay.cc

```diff
@@ -0,0 +1,938 @@
+
+    if ((p == "jerasure") || (p == "isal")) {
+      mds.profile["plugin"] = p;
+      pft.profile["plugin"] = p;
+    }
+    else {
+      mds.profile["plugin"] = "jerasure";
+      pft.profile["plugin"] = "jerasure";
+      *ss << "scalar_mds " << p << "is not currently supported using the default reed_sol_van" << std::endl;
+    }
+  }
+
+  if (profile.find("technique") == profile.end() ||
+      profile.find("technique")->second.size() == 0) {
+    mds.profile["technique"] = "reed_sol_van";
+    pft.profile["technique"] = "reed_sol_van";
+    *ss << "technique not found in profile picking default reed_sol_van" << std::endl;
+  } else {
+    std::string p = profile.find("technique")->second;
+    *ss << "recieved technique as " << p << std::endl;
+    mds.profile["technique"] = p;
+    pft.profile["technique"] = p;
+  }
+  
+  if ((d < k) || (d > k+m-1)) {
+    *ss << "value of d " << d
+        << " must be within [ " << k << "," << k+m-1 << "]" << std::endl;
+    err = -EINVAL;
+    return err;
+  }
+  
+  q = d-k+1;
+  if ((k+m)%q) {
+    nu = q - (k+m)%q;
+    *ss << "Clay: (k+m)%q=" << (k+m)%q
+	<< "q doesn't divide k+m, to use shortening" << std::endl;
+  } else {
+    nu = 0;
+  }
+
+  if (k+m+nu > 254) {
+    err = -EINVAL;
+    return err;
+  }
+  
+  mds.profile["k"] = std::to_string(k+nu);
+  mds.profile["m"] = std::to_string(m);
+  mds.profile["w"] = '8';
+  
+  pft.profile["k"] = '2';
+  pft.profile["m"] = '2';
+  pft.profile["w"] = '8';
+  
+  dout(10) << __func__ << " k:" << k << " m: " << m << " w:" << w << dendl;
+    
+  t = (k+m+nu)/q;
+  sub_chunk_no = pow_int(q, t); 
+
+  return 0;
+}
+
+int ErasureCodeClay::get_repair_sub_chunk_count(const set<int> &want_to_read)
+{
+  int repair_subchunks_count = 1;
+
+  int weight_vector[t];
+  memset(weight_vector, 0, t*sizeof(int));
+
+  for (set<int>::iterator i = want_to_read.begin();
+       i != want_to_read.end(); ++i) {
+    weight_vector[(*i)/q]++;
+  }
+
+  for (int y = 0; y < t; y++) repair_subchunks_count = repair_subchunks_count*(q-weight_vector[y]);
+
+  return sub_chunk_no - repair_subchunks_count;
+}
+
+int ErasureCodeClay::is_repair(const set<int> &want_to_read,
+			       const set<int> &available_chunks) {
+
+  if (includes(
+	       available_chunks.begin(), available_chunks.end(), want_to_read.begin(), want_to_read.end())) return 0;
+
+  if (want_to_read.size() > 1) return 0;
+
+  for (set<int>::iterator i=want_to_read.begin();
+       i != want_to_read.end(); ++i) {
+    int lost_node_id = (*i < k) ? *i: *i+nu;
+    for (int x = 0; x < q; x++) {
+      int node = (lost_node_id/q)*q+x;
+      node = (node < k) ? node : node-nu;
+      if (want_to_read.find(node) == want_to_read.end()) {//node from same group not erased
+        if (available_chunks.find(node) == available_chunks.end()) return 0;//node from same group not available as well
+      }
+    }
+  }
+
+  if (available_chunks.size() < (unsigned)d) return 0;
```

Comment:
In the appendix of 'Clay Codes: Moulding MDS Codes to Yield an MSR Code', it's claimed that when d=n-1 the it can recover from up to (q-1) failures  that occur within a single y-section. The example in appendix allows 12 available nodes, less than d=13, to repair two failures (0,0) and (2,0). So why when available_chunks.size() < d then return 0?

Reply:
That is true. I have currently kept the is_repair simple to support single chunk failure case alone. Planning to add multiple chunk failures cases in the next step. (The backup code for multiple failure correction can be found at https://github.com/mynaramana/ceph/tree/multiple_erasures)